Handling of multiple attribute values and LDAP entries for the 'attribute' query.

Ayanda-D opened the pull request:
Fixes: #16

Ayanda-D committed:
Adds handling of multiple attribute values, and
multiple LDAP entries for the 'attribute' query.

Ref: #16

michaelklishin commented:
@Ayanda-D any chance we can add an integration test? [Our test setup](https://github.com/rabbitmq/rabbitmq-auth-backend-ldap/blob/master/README-tests.md) isn't great but we do have a test suite.

Ayanda-D commented:
@michaelklishin sure, I'm having a look.

michaelklishin commented:
@Dzol this primarily needs tests.

Ayanda-D committed:
Adds 'memberOf' attribute overlay configuration.

Ref: #16.

Ayanda-D committed:
Adds 'staff' LDAP testing group.

Ref: #16.

Ayanda-D committed:
Adds posix test user accounts, 'Carol' and 'Peter'.
Adds test user 'Peter' to all test groups.

Ref: #16.

Ayanda-D committed:
Updates LDAP 'match' query to carry out a bidirectional match, in
case the REQuery returned multiple attributes, i.e. RE strings.

Adds a format_multi_attr/3 match all clause.

Ref: #16.

Ayanda-D committed:
Adds success and failure tests for handling multiple attributes,
using the 'memberOf' attribute.

Adds filter lists per 'login' case.

Cosmetics.

Ref: #16.

Ayanda-D commented:
Tests for #16 are now covered.

michaelklishin commented:
@Ayanda-D thank you. This no longer merges into `stable`, please resolve conflicts.

Ayanda-D commented:
@michaelklishin sorted.

michaelklishin commented:
Not assigning anyone as this will be reviewed by @dcorbacho, @hairyhum, and me.

michaelklishin commented:
Seeding fails for me on this branch (but works on `stable`):

```
# ./example/setup.sh
#
# [snip]
SASL/EXTERNAL authentication started
SASL username: gidNumber=0+uidNumber=0,cn=peercred,cn=external,cn=auth
SASL SSF: 0
adding new entry "cn=module,cn=config"

adding new entry "olcDatabase=bdb,cn=config"

adding new entry "cn=module,cn=config"

adding new entry "olcOverlay={0}memberof,olcDatabase={1}hdb,cn=config"
ldap_add: No such object (32)
    matched DN: cn=config
```

dcorbacho commented:
I get something different after run setup.sh and seed.sh, but the tests pass afterwards:

```
$ sh seed.sh 
SASL/EXTERNAL authentication started
SASL username: gidNumber=0+uidNumber=0,cn=peercred,cn=external,cn=auth
SASL SSF: 0
adding new entry "cn=module,cn=config"
ldap_add: Other (e.g., implementation specific) error (80)
        additional info: <olcModuleLoad> handler exited with 1

adding new entry "cn=module,cn=config"
ldap_add: Other (e.g., implementation specific) error (80)
        additional info: <olcModuleLoad> handler exited with 1

modifying entry "cn=module{1},cn=config"
ldap_modify: Type or value exists (20)
        additional info: modify/add: olcModuleLoad: value #0 already exists

adding new entry "olcOverlay={1}refint,olcDatabase={1}hdb,cn=config"

adding new entry "dc=example,dc=com"
ldap_add: Already exists (68)

adding new entry "ou=groups,dc=example,dc=com"
ldap_add: Already exists (68)

adding new entry "ou=vhosts,dc=example,dc=com"
ldap_add: Already exists (68)
```

dcorbacho commented on src/rabbit_auth_backend_ldap.erl:
```diff
@@ -323,6 +323,36 @@ get_or_create_conn(IsAnon, Servers, Opts) ->
+    format_multi_attr(io_lib:printable_list(Attrs), Attrs, []).
+
+format_multi_attr(true, Attrs, _Acc)       -> Attrs;
+format_multi_attr(_, [], Acc)              -> lists:flatten(Acc);
```
Maybe all the `false` case could be replaced by `string:join(Attrs, "; ")` ?

Ayanda-D replied:
@dcorbacho yes they could. I'll update this. Thanks

dcorbacho commented on src/rabbit_auth_backend_ldap.erl:
```diff
@@ -323,6 +323,36 @@ get_or_create_conn(IsAnon, Servers, Opts) ->
+  format_multi_attr(F, Rem, [Acc | ["; ", H]]).
+
+%% In case of multiple attributes, check for equality bi-directionally
+is_multi_attr_member(Str1, Str2) ->
```
Equivalent to `lists:member(Str1, Str2) orelse lists:member(Str2, Str1)`

Ayanda-D replied:
@dcorbacho true. I'll update. Thanks for this :+1:

hairyhum commented:
This PR is for `stable`, but contains `.schema` file. Looks like some leak from `master`.

hairyhum commented:
@Ayanda-D according to commit history this PR is based on `master` and needs to be rebased or cherry-picked to be based `stable`.

Ayanda-D commented:
@hairyhum hmm, okay cool, noted. I'll resolve this. thanks :+1:

Ayanda-D committed:
Updates multi atrributes formatting to use string:join/2.
Updates is_multi_attr_member/2 to use a short-circuit expression.

Ref: #16.

Ayanda-D committed:
Cosmetics

Ayanda-D commented:
Done with the suggested updates.

dcorbacho commented:
@Ayanda-D The commit history still contains the .schema changes, could you delete them?

hairyhum commented:
Looks fine to me.

michaelklishin commented:
I cannot run the setup/seed script on this branch:

```
SASL/EXTERNAL authentication started
SASL username: gidNumber=0+uidNumber=0,cn=peercred,cn=external,cn=auth
SASL SSF: 0
adding new entry "cn=module,cn=config"

adding new entry "olcDatabase=bdb,cn=config"

adding new entry "cn=module,cn=config"

adding new entry "olcOverlay={0}memberof,olcDatabase={1}hdb,cn=config"
ldap_add: No such object (32)
    matched DN: cn=config
```

while it's OK on `stable`.

Ayanda-D commented:
@michaelklishin this works just fine for me, & considering parent entries of `olcOverlay={0}memberof,olcDatabase={1}hdb,cn=config` are successfully added. I'm not reproducing this same error you're getting.

dcorbacho commented:
I'm getting this:

```
ubuntu@ubuntu:~/rabbitmq-public-umbrella/deps/rabbitmq_auth_backend_ldap/example$ ./seed.sh 
SASL/EXTERNAL authentication started
SASL username: gidNumber=0+uidNumber=0,cn=peercred,cn=external,cn=auth
SASL SSF: 0
adding new entry "cn=module,cn=config"
ldap_add: Other (e.g., implementation specific) error (80)
        additional info: <olcModuleLoad> handler exited with 1
```

Ayanda-D commented:
@dcorbacho `./seed.sh` should already be executed as part of `./setup.sh`. no need to run it independently.

dcorbacho commented:
@Ayanda-D right, then it works for me.

hairyhum commented:
`./setup.sh` works fine for me on fresh installed `slapd-2.4.31-1` on Ubuntu 14.04.3 LTS

michaelklishin commented:
OK, then the error I'm running into shouldn't prevent this from being merged.
